[MRG+1] Isotonic regression duplicate fixes

amueller opened the pull request:
Fixes #4184. With tests from #4185.
This is a stupid pure-python version. I'm not sure if there is an easy way to vectorize.
It now implements the "secondary" method, which basically replaces duplicate points with weighted averages. This is the only method that makes `fit_transform` behave identical to `fit().transform()`. I used a naive implementation of `fit_transform()`.

mjbommar committed:
Adding unit test to cover ties/duplicate x values in Isotonic Regression re: issue #4184

Expanding tests to include ties at both x_min and x_max

Updating unit test to include reference data against R's isotone gpava() with ties=primary

Adding R and isotone package versions for reproducibility/documentation

Removing double space in docstring

Combining tests for fit and transform with ties; fixing spelling error

amueller commented:
Ok so there is some fun here.
The test ``test_isotonic_regression`` tests that ``# check it doesn't change y when all x are equal``.
Which is only possible with primary or tertiary method. Obviously if you don't change y when all x are equal, you can not have ``transform`` and ``fit_transform`` have the same results.

What do we do?

amueller commented:
To be clear, we have two choices:
1) Remove the test, breaking backward compatibility with (somewhat broken) behavior.
2) Implement primary (or ternary) method, making ``fit(X).transform(X)`` not equal to ``fit_transform(X)``.

agramfort commented:
travis is not happy

agramfort commented:
2) is even more evil than 1)

amueller commented:
Well currently we have tests that can not be fulfilled simultaneously. We test that ``fit_transform`` doesn't change ``y`` if all ``X`` are identical, and we have tests that ensure ``fit().transform()`` and ``fit_transform()`` are equivalent.
If we agree that 1) is the less evil way, I'll remove the other test. I'm currently writing the cython that can replace my naive python.

agramfort commented:
1) sounds good to me

mjbommar commented:
Do we need to apply the ``_make_unique`` transform prior to the determination of ``increasing`` as well?

mjbommar commented:
@amueller , regarding a slightly faster way to implement the code, how about the comprehension bellow for non-weighted:
```
unique_data = [(x, np.average(y[X==x])) for x in np.unique(X)]
x_unique, y_unique = zip(*unique_data)
```

And for weighted:
```
unique_data = [(x, np.average(y[X==x], sample_weights[X==x])) for x in np.unique(X)]
x_unique, y_unique = zip(*unique_data)
```

amueller commented:
@mjbommar just checked in cython to do it. It might not be pretty, but it is pretty fast ;)

amueller commented:
Yeah I think the missing order was a bug and needs a regression test.
Otherwise the PR seems good to go.

amueller commented:
@mjbommar :+1:
> don't we need to apply the _make_unique transform prior to the determination of increasing as well? 
where? Do we?

mjbommar commented:
When ``check_increasing`` runs in ``_build_y``, we pass it data without running ``_make_unique``.  Especially if someone were to choose ``"pearson"``, you could imagine outliers that were averaged out that could switch the direction.

On the flip side, the p-values of the "true" relationship are more informative than the p-values for the x-averaged data.t

amueller commented:
I feel that if we include the ``_make_unique`` in the ``check_increasing`` test, the behavior might be harder to understand. You are right, it might change behavior in edge-cases. How strongly do you feel about this?

mjbommar commented:
@amueller , not too strong, to be honest.  What about adding an ``"auto_unique"`` option to the ``increasing`` argument options?

ogrisel commented on sklearn/_isotonic.pyx:
```diff
@@ -69,3 +69,50 @@ def _isotonic_regression(np.ndarray[DOUBLE, ndim=1] y,
+@cython.cdivision(True)
+def _make_unique(np.ndarray[dtype=np.float64_t] X,
+                  np.ndarray[dtype=np.float64_t] y,
+                  np.ndarray[dtype=np.float64_t] sample_weights):
```
Would be good to add a docstring to explain that the number of samples is reduced whenever there is a tie in X: the matching y values are averaged using the weights and the sample_weights are accumulated to conserve the original info.

It is my understanding that X should be sorted before calling this utility, this should also be made explicit in the docstring or in an inline comment to improve understandability of the code.

amueller replied:
done.

ogrisel commented on sklearn/_isotonic.pyx:
```diff
@@ -69,3 +69,50 @@ def _isotonic_regression(np.ndarray[DOUBLE, ndim=1] y,
+    cdef np.ndarray[dtype=np.float64_t] x_out = np.empty(unique_values)
+    cdef np.ndarray[dtype=np.float64_t] weights_out = np.empty(unique_values)
+
+
```
cosmetics: only 1 blank line here.

ogrisel commented on sklearn/_isotonic.pyx:
```diff
@@ -69,3 +69,50 @@ def _isotonic_regression(np.ndarray[DOUBLE, ndim=1] y,
+            current_y = y[j] * sample_weights[j]
+            current_count = 1
+        else:
+            current_weight +=sample_weights[j]
```
missing whitespace

amueller commented on sklearn/_isotonic.pyx:
```diff
@@ -69,3 +69,50 @@ def _isotonic_regression(np.ndarray[DOUBLE, ndim=1] y,
+def _make_unique(np.ndarray[dtype=np.float64_t] X,
+                  np.ndarray[dtype=np.float64_t] y,
+                  np.ndarray[dtype=np.float64_t] sample_weights):
+    # assumes ordered X
```
There is an inline comment here, but I should probably write a proper docstring. Wil ldo that now.

ogrisel commented on sklearn/isotonic.py:
```diff
@@ -249,8 +265,13 @@ def _build_y(self, X, y, sample_weight):
+        if sample_weight is None:
+            sample_weight = np.ones(len(y))
+        unique_X, unique_y, unique_sample_weight = _make_unique(
+            X[order].astype(np.float64), y[order].astype(np.float64),
```
`.astype` will cause an additional memory copy, better use `from sklearn.utils.fixes import astype`.

amueller replied:
done

ogrisel commented on sklearn/tests/test_isotonic.py:
```diff
@@ -76,9 +91,67 @@ def test_isotonic_regression():
+     Isotone Optimization in R: Pool-Adjacent-Violators Algorithm
+    (PAVA) and Active Set Methods
+    """
+
```
Why this hole in the docstring?

amueller replied:
probably rebase error. fixed.

amueller committed:
Implement "secondary" tie strategy in isotonic.

This strategy allows us to make fit_transform(X) behave the same as
fit(X).transform(X)
Remove test for not touching duplicate entries in fit_transform().

ogrisel committed:
ENH no need for tie breaking jitter in calibration

The isotonic regression routine now implements deterministic
tie-breaking by default.

ogrisel commented:
This LGTM. All the calibration examples still look good. I have not checked the isotonic regression examples. I opened a PR against this PR https://github.com/amueller/scikit-learn/pull/26 to remove the calibration random jitter used for tie breaking.

ogrisel commented:
@amueller before merging, please rewrite the commit message to make it more descriptive of the actual content of this commit.

amueller commented:
merged your PR. We need another review, right?

ogrisel commented:
Maybe @agramfort is around for a final review. @mjbommar WDYT of the current state of this PR?

amueller commented:
reworded the commit message (rebase -i ftw).

mjbommar commented:
Looked good to me earlier this week.  Commits today were mostly cosmits, right?

I have a fix for issue #4297 ready to PR once this is merged too, which would result in passing tests.

amueller commented:
Yup, only cosmetics here.

amueller commented:
This is kind of an important bug-fix for the release so it would be great if any other dev had time for a look....

mjbommar commented:
And after this is merged, we need to pull my followup for the infinite loop too, so another PR review (though brief) after.

ogrisel commented:
@jnothman it would be great if you could have a look at this one.

amueller commented:
Or maybe @GaelVaroquaux ;)

ogrisel commented:
@agramfort if you have time tomorrow, I would really like to have this fix in for the 0.16 beta but we cannot delay the release of the beta further.

mjbommar commented:
@amueller , would swapping in my comprehensions make this many fewer LOC to review?  ``%timeit`` gave me <50ms timings for realistic sample sizes, though I didn't actually compare your cython to it.

amueller commented:
@mjbommar really? It gave me 40s while mine was three or four orders of magnitude faster... Also, I'm not sure the Cython is the problem. I'll check how many samples I used when I'm on the other box. But I think it was around 1e6 or 1e7.

amueller commented:
Btw, I assumed that all samples but one are distinct. I think it is reasonable to assume that most samples are distinct. In this case, the list comprehension has quadratic complexity, right?

mjbommar commented:
@amueller , not arguing your solution was better, just really hoping we can get this into the 0.16 release and wondering what would help the rest of the team review

amueller commented:
Yeah I understand your motivation, I just wanted to say that I think there is good reason. We will get this into 0.16. I suspect the reason that it is not reviewed yet is that it takes people a bit of time to understand what the actual problem was.

ogrisel commented:
Alright, let's merge this fix.
